src/notifications/NotificationsManager.ts

```diff
@@ -73,37 +94,174 @@ class NotificationsManager {
   protected db: DB;
   protected keyRing: KeyRing;
   protected nodeManager: NodeManager;
+  protected taskManager: TaskManager;
   protected messageCap: number;
 
   /**
-   * Top level stores MESSAGE_COUNT_KEY -> number (of messages)
+   * Map for hashing NotificationIds to find TaskIds.
+   * The key must be NotificationIdEncoded, so that the string representation is hashed rather than the reference to the NotificationId object.
    */
-  protected notificationsDbPath: LevelPath = [this.constructor.name];
+  protected outboxNotificationIdEncodedToTaskIdMap: Map<
+    NotificationIdEncoded,
+    TaskId
+  >;
+
   /**
-   * Messages stores NotificationId -> string (message)
+   * Top level stores inbox/outbox -> Notifications, numMessages
    */
-  protected notificationsMessagesDbPath: LevelPath = [
-    this.constructor.name,
-    'messages',
+  protected notificationsManagerDbPath: LevelPath = [this.constructor.name];
+  protected notificationsManagerInboxDbPath: LevelPath = [
+    ...this.notificationsManagerDbPath,
+    'inbox',
+  ];
+  protected notificationsManagerOutboxDbPath: LevelPath = [
+    ...this.notificationsManagerDbPath,
+    'outbox',
   ];
-  protected notificationsMessageCounterDbPath: KeyPath = [
-    ...this.notificationsDbPath,
-    MESSAGE_COUNT_KEY,
+  /**
+   * Notifications stores NotificationId -> Notification
+   */
+  protected notificationsManagerInboxNotificationsDbPath: LevelPath = [
+    ...this.notificationsManagerInboxDbPath,
+    'notifications',
   ];
+  protected notificationsManagerInboxNotificationsCounterDbPath: KeyPath = [
+    ...this.notificationsManagerInboxDbPath,
+    NOTIFICATIONS_COUNT_KEY,
+  ];
+  protected notificationsManagerOutboxNotificationsDbPath: LevelPath = [
+    ...this.notificationsManagerOutboxDbPath,
+    'notifications',
+  ];
+  protected notificationsManagerOutboxNotificationsCounterDbPath: KeyPath = [
+    ...this.notificationsManagerOutboxDbPath,
+    NOTIFICATIONS_COUNT_KEY,
+  ];
+
+  protected inboxNotificationIdGenerator: () => NotificationId;
+  protected outboxNotificationIdGenerator: () => NotificationId;
 
-  protected notificationIdGenerator: () => NotificationId;
+  public readonly sendNotificationHandlerId =
+    `${this.constructor.name}.sendNotificationHandler` as TaskHandlerId;
+
+  protected sendNotificationHandler: TaskHandler = async (
+    _ctx,
+    _taskInfo,
+    {
+      nodeIdEncoded,
+      notificationIdEncoded,
+      retries,
+      attempt,
+    }: {
+      nodeIdEncoded: NodeIdEncoded;
+      notificationIdEncoded: NotificationIdEncoded;
+      retries: number;
+      attempt: number;
```

Comment:
Unless we really need to know the total attempts that were allowed. We only really need to count down the attempts left and track 1 number for the logic.

So rather than count up till the retry limit, start at the retry limit and count to 0.

Reply:
i tried to do that originally, but the attempts are needed to keep track of the exponential timeout

Reply:
You can just double the previous delay. Unless you can't access the task info inside of the handler?

Reply:
i've made it so that the delay is doubled instead